docs/features/readonly-variables-and-parameters.md

```diff
@@ -0,0 +1,61 @@
+### High Level Language Proposal:  
+Allow 'readonly' to be used on local variables and on parameters. 
+
+'readonly' on local variables and 'parameters' will work similarly as it does for 'fields' (with important differences listed in the next paragraph).  Namely, 'readonly' local variables and parameters cannot be written to in any way (i.e. through assignment, use of ```++/+=```, passing as a 'ref/out', etc.). 
+
+Important differences:
+unlike 'readonly' fields, 'readonly' locals must have an initializer.
+unlike 'readonly' fields (which can be written to multiple times in constructors), 'readonly' locals can only have the initializer and can never be written to again.
+unlike 'readonly' fields, the 'readonly'ness is not visible at the IL level.  'readonly' is a source-level-only attribute which indicates how the local and parameter can be treated inside a method body.  IN this way it is similar to 'async'.  The modifier appears on the signature, but it only affects the code of the body.
+
+Allowed locations:
+1. Method parameters (non-abstract/extern)  
+2. Constructor parameters  
+3. Indexer parameters  
+4. Operator parameters  
+5. Lambda parameters  
+6. Anonymous method parameters  
+7. Local function parameters  
+
+
+Disallowed locations:  
+1. Delegate parameters  
+2. Methods/indexers witout bodies (i.e. extern/abstract/interface)
+
+
+Interesting questions:  
+1. Allow "readonly this" (for extension methods)?  Current implementation allows it.  I don't see why not to allow it.  
+2. Allow "readonly params"?  Current implementation allows it.  I don't see why not to allow it.  
+3. Allow "readonly ref"?  Currently no.  My understanding is that there are design/perf concerns here.  So i'm just disallowing it for now.  
+4. Allow "(readonly i) => ..." (i.e. a readonly *implicit* lambda parameter).  Current implementation allows it.  
```

Comment:
@CyrusNajmabadi (1) Since this is allowed, woudn't it be nice to also allow `readonly i` instead of `readonly var i` for locals? (same question applies to local constants — I'm thinking that since `from var item` is not allowed in a query expression, or `let var` is not allowed either and a optional type is permitted (from pattern spec draft), it'll be more consistent without `var`. Just my opinion. (2) Does it make sense to also have `const` parameters? meaning that the parameter must come from a compile-time literal or another `const`.